### PR TITLE
Senghoung/edit review feature

### LINKED ISSUE #273: Add confirmation when deleting review on reviews page, and ability to update review
- [ ] Delete review confirmation
- [ ] Update review

### PR DESCRIPTION
**Pull request by SenghoungLim**
## Description
- Added edit review feature in user's review menu/tab.
- Added delete confirmation dialog for user's review.
- Added user-friendly design of the edit pen icon accommodating dark mode.

## Screenshots
- Showcase of delete review & pen UI for dark/light mode
Showcase of delete review & pen UI for dark/light mode
- Showcase delete review confirmation & dark/light mode UI
Showcase delete review confirmation & dark/light mode UI

## Steps to verify/test this change:
- [ ] Verify changes work as expected on staging instance

## Final Checks:
- [ ] Verify successful deployment

(optional)
- [ ] Write tests
- [ ] Write documentation

## Issues

Closes #273

### CODE REVIEW

**Comment by Awesome-E** [comment]
Not going to review the whole PR since I'm not assigned to it, but I do have a few comments on design

I know what you have is the default style for the modal, but I think you should make it look more like the design and/or the one I made in #419. To be more specific, that is:

- Making the X bigger to match the size of the heading on the left
- not having the borders on bottom and top because that looks cleaner
- Modal background color should be one of the overlay colors (I forget which one, but it's in my PR)

---

Also, the edit button from your GIF appears to not be square, which makes the backdrop an oval instead of a circle. To be honest, I think don't think we should have it as a circle at all. Instead, I think we should keep the UI consistent with other parts of PPC and do the same thing like we do in the roadmap – that is, have a transparent button that is of the variant "light/dark depending on theme".

File: package.json [hunk 1]
```diff
@@ -10,6 +10,7 @@
   },
   "dependencies": {
     "dotenv-flow": "^4.0.1",
+    "react-icons": "^5.0.1",
```
Comment (js0mmer) [inline package.json:13]: Would prefer not to add another icon pack/dependency just for one icon. We already have react bootstrap icons, let's use [one of their icons.](https://icons.getbootstrap.com/?q=pen)

File: site/src/component/Review/SubReview.tsx [hunk 2]
```diff
@@ -91,6 +96,11 @@ const SubReview: FC<SubReviewProps> = ({ review, course, professor, colors, colo
 
   return (
     <div className="subreview">
+      {editable && editReview && (
+        <div className="edit-pen-icon" onClick={() => editReview(review, course, professor)}>
+          <FaPen />
+        </div>
```
Comment (js0mmer) [inline site/src/component/Review/SubReview.tsx:102]: Should make this a button not a div for accessibility purposes and then style it appropriately. I agree with Ethan's comments on the button styling.

File: site/src/component/ReviewForm/ReviewForm.tsx [hunk 3]
```diff
@@ -81,19 +82,48 @@ const ReviewForm: FC<ReviewFormProps> = (props) => {
...
+      setContent(props.review?.reviewContent);
+      setSelectedTags(props.review?.tags);
+      setAttendance(props.review?.attendance);
+      setTakeAgain(props.review?.takeAgain);
+      setTextbook(props.review?.textbook);
+      setUserName(props.review?.userDisplay);
+      setProfessor(props.review?.professorID);
+      setCourse(props.review?.courseID);
+    }
+  }, [showForm]);
```
Comment (js0mmer) [inline site/src/component/ReviewForm/ReviewForm.tsx:107]: Add the dependencies to clear the warning. The lint check will allow warnings but our goal should be to have none so let's make sure not to add anymore.

```suggestion
  }, [showForm, cookies.user, props]);
```

File: site/src/component/ReviewForm/ReviewForm.tsx [hunk 4]
```diff
@@ -115,35 +145,69 @@ const ReviewForm: FC<ReviewFormProps> = (props) => {
...
-      score: 0,
-      takeAgain: takeAgain,
-      textbook: textbook,
-      attendance: attendance,
-      tags: selectedTags,
-      captchaToken: captchaToken,
-    };
-    if (content.length > 500) {
-      setOverCharLimit(true);
+    if (props.editable === false) {
```
Comment (js0mmer) [inline site/src/component/ReviewForm/ReviewForm.tsx:148]:
```suggestion
    if (!props.editable) {
```

File: site/src/component/UserReviews/UserReviews.tsx [hunk 5]
```diff
@@ -22,9 +37,32 @@ const UserReviews: FC = () => {
...
+  //Edit Review
+  const editReview = (review: ReviewData, course?: CourseGQLData, professor?: ProfessorGQLData) => {
+    setCourseToEdit(course);
+    setProfessorToEdit(professor);
+    setReviewToEdit(review);
+    console.log('Data received from SubReview:', {
+      course,
+      professor,
+      review,
+    });
```
Comment (js0mmer) [inline site/src/component/UserReviews/UserReviews.tsx:56]:
```suggestion
```

File: site/src/component/UserReviews/UserReviews.tsx [hunk 6]
```diff
@@ -22,9 +37,32 @@ const UserReviews: FC = () => {
...
+    dispatch(setFormStatus(true));
+    document.body.style.overflow = 'hidden';
+    console.log('Edit Review clicked!');
```
Comment (js0mmer) [inline site/src/component/UserReviews/UserReviews.tsx:59]:
```suggestion
```

File: site/src/component/UserReviews/UserReviews.tsx [hunk 7]
```diff
@@ -39,14 +77,45 @@ const UserReviews: FC = () => {
...
+                  <Modal.Header closeButton>
+                    <Modal.Title>Delete Review Confirmation</Modal.Title>
+                  </Modal.Header>
+                  <Modal.Body>
+                    Deleting a review will remove it permanently. Are you sure you want to proceed?
+                  </Modal.Body>
+                  <Modal.Footer>
+                    <Button variant="secondary" onClick={handleClose}>
+                      Close
+                    </Button>
```
Comment (js0mmer) [inline site/src/component/UserReviews/UserReviews.tsx:102]: I would say cancel instead.

```suggestion
                    <Button variant="secondary" onClick={handleClose}>
                      Cancel
                    </Button>
```

**Review by js0mmer — CHANGES_REQUESTED** [review]
Works well except for 2 things:
* reCAPTCHA has no effect, should implement or just omit it for review edits which I think may be safe to do, given that the resource already exists
* Form validation does not reset. If I edit a review then go to edit it again, green outlines and check marks are still there.
image

Left additional comments, mainly on small refactors and design

**Comment by js0mmer** [comment]
replaced by #487

**Pull request closed without merging**